File: aizu/tyama_aizuALDS1~2D.py
#!/usr/bin/python
import sys
if sys.version_info[0]>=3: raw_input=input

def insertionSort(a,g):
	global cnt
	for i in range(g,len(a)):
		v = a[i]
		j = i - g
		while j >= 0 and a[j] > v:
			a[j+g] = a[j]
			j = j - g
			cnt+=1
			a[j+g] = v

def shellSort(a):
	global cnt
	cnt = 0
	g = []
	h = 1
	while h <= len(a):
		g.append(h)
		h = 3*h+1
	g.reverse()
	m = len(g)
	print(m)
	print(' '.join(map(str,g)))
	for i in range(m):
		insertionSort(a,g[i])

a=[int(raw_input()) for i in range(int(raw_input()))]
shellSort(a)
print(cnt)
for e in a: print(e)

File: aizu/tyama_aizuALDS1~2D.rb
#!/usr/bin/ruby
def insertionSort(a,g)
	for i in g..a.size-1
		v = a[i]
		j = i - g
		while j >= 0 && a[j] > v
			a[j+g] = a[j]
			j = j - g
			$cnt+=1
			a[j+g] = v
		end
	end
end

def shellSort(a)
	$cnt = 0
	g = []
	h = 1
	while h <= a.size
		g << h
		h = 3*h+1
	end
	g.reverse!
	m = g.size
	p m
	puts g*' '
	for i in 0..m-1
		insertionSort(a,g[i])
	end
end

a=gets.to_i.times.map{gets.to_i}
shellSort(a)
p $cnt
puts a